content/2018/swalladge-vim-and-git.md

```diff
@@ -1,197 +1,320 @@
+configuration below (this is my config with Neovim; launching Vim in diffmode is
+slightly different):
 
-If checking out different versions of a file in Git while that file, Vim will
-complain about the file being changed externally and prompt for an action to
-take. With `:set autoread`, external changes will be automatically loaded into
-the buffer for a smoother workflow. `u` will undo the change.
+```
+[merge]
+    tool = nvimdiff4
+    # if not using a tool name with builtin support, must supply mergetool cmd
+    # as below
 
+[mergetool "nvimdiff4"]
+    cmd = nvim -d $LOCAL $BASE $REMOTE $MERGED -c '$wincmd w' -c 'wincmd J'
 
-If you don't want to leave vim, it is simple to run a git command by shelling
-out. For example: `:!git status`. `%` expands to the current filename, so
-something like `:!git log %` will show the commit history of the file open in
-the current buffer.
+[mergetool "nfugitive"]
+    cmd = nvim -f -c "Gdiff" "$MERGED"
+```
 
-New Vim and Neovim versions also have an integrated terminal. `:terminal`
+The first method requires no plugins. It is simply Vim (Neovim) launched in
+diffmode with the 4 files Git provides in environment variables. The wincmds
+executed move the working file to full-width along the lower half of the window.
 
-TODO
+The second method uses the vim-fugitive plugin to automatically set up the
+layout (it will choose horizontal or vertical splits depending on the terminal
+size. It still uses Vim's diffmode, but only shows 3 files (local, remote, and
+index).
 
-### Tmux
+For both methods, use `diffget` and `diffput` commands (with corresponding `do`
+and `dp` bindings) to resolve conflicts.  Something I find helpful is to include
+the buffer number in the status line as a quick reference when giving the buffer
+number to `do` or `dp`. The item `%n` will do that. See `:h 'statusline'` for
+more.
 
-- popular
-- switch between tmux panes/windows; one for vim, another for shell/git
-- no suspend/fg required
+For further information, I recommend the [vimcasts screencast on resolving merge
+conflicts][vimcasts-vimdiff].
 
+Likewise, git `difftool` can be set to use Vim too for displaying diffs:
+
+```
+[diff]
+    tool = nvimdiff2
+
+[difftool "nvimdiff2"]
+    cmd = nvim -d $LOCAL $REMOTE
+```
 
-### Plugins
 
-- plugins can augment the above
-- plugins can replace the above
+### Interactive rebasing
 
+I haven't had to perform rebases very often and so can't speak authoritatively
+on the subject. I need to include it here though for completeness.
 
-## Let's talk about plugins
+I do know that Vim is perfectly suited to the task though. Common actions map
+neatly to Vim idioms: change a command (`ciw`), remove a commit (`dd`), or
+reorder lines (`<count>dd`, move up/down, `p` or `P`). If you find yourself
+repeating a certain action often, it is simple to create a new keybind for it as
+in [this Reddit post][reddit-rebase-bindings].
```

Comment:
Perhaps worth mention that this is built into Vim already by a `Cycle` command:

https://github.com/tpope/vim-git/blob/master/ftplugin/gitrebase.vim

This inspired me to make `<C-x/a>` switch between these, too:

https://github.com/Konfekt/vim-CtrlXA/commit/48167abfebb2d7ed1652d440fbfc664afdcdc053

Reply:
Wow I never realized that was shipped with Vim! :open_mouth: Thanks for the tip. :)